library/choose_disk.py

```diff
@@ -0,0 +1,651 @@
+    '''
+    ceph_disks = []
+    non_ceph_disks = []
+
+    for physical_disk in sorted(physical_disks):
+        if "ceph_prepared" in physical_disks[physical_disk]:
+            # We shall only return the ceph disks from the same type
+            # If we search for journals, don't return data disks :
+            #   Note that we don't neither them into 'non_ceph_disks' as they are not free
+            pdisk = dict(physical_disks[physical_disk])
+
+            if expected_type in pdisk["ceph_prepared"]:
+                logger.debug("get_keys_by_ceph_order: Keeping %s", physical_disk)
+                ceph_disks.append(physical_disk)
+            else:
+                logger.debug("get_keys_by_ceph_order: %s doesn't have the proper ceph type", physical_disk)
+        else:
+            non_ceph_disks.append(physical_disk)
+
+    return ceph_disks + non_ceph_disks
+
+
+def find_match(physical_disks, lookup_disks, module=None):
+    '''
+    Find a set of matching devices in physical_disks
+    '''
+    matched_devices = {}
+    ignored_devices = []
+
+    # associate a keyword and the associated function
+    OPERATORS = {
+        "=": _equal,
+        "equal": _equal,
+        "gt": _gt,
+        "gte": _gte,
+        "lt": _lt,
+        "lte": _lte,
+        "and": _and,
+    }
+
+    logger.info("Looking for matches")
+
+    # Inspecting every disk we search for
+    for disk in sorted(lookup_disks):
+
+        infinite = False
+        current_lookup = dict(lookup_disks[disk])
+        current_type = ""
+
+        # Does the user want _all_ the devices matching that description ?
+        if "infinite" in current_lookup:
+            infinite = True
+            del current_lookup["infinite"]
+
+        # We cannot keep the disk type as a feature to lookup unless the
+        # matching with a real device would fail
+        current_type = current_lookup["ceph_type"]
+        del current_lookup["ceph_type"]
+
+        if len(ignored_devices) == len(physical_disks):
+            logger.info(" Skipping %s as no more free devices to match", disk)
+            continue
+
+        logger.info(" Inspecting %s", disk)
+        # Trying to find a match against all physical disks we have
+        for physical_disk in get_keys_by_ceph_order(physical_disks, current_type):
+            # Avoid reusing an already matched physical disk
+            if physical_disk in ignored_devices:
+                continue
+
+            current_physical_disk = physical_disks[physical_disk]
+            match_count = 0
+            # Checking what features are matching
+            for feature in current_lookup:
+                if feature not in current_physical_disk:
+                    continue
+
+                # Assign left and right operands
+                right = current_lookup[feature]
+                left = current_physical_disk[feature]
+
+                def evaluate_operator(left, right):
```

Comment:
doesn't seem like this function should be nested in here, by making it a closure it is making it harder to unit test (the inputs are currently forced to be passed onto the parent function). It is better to have it outside so that testing is improved.